cpp/src/arrow/compute/kernels/vector_dictionary.cc

```diff
@@ -0,0 +1,201 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "arrow/array/array_base.h"
+#include "arrow/array/array_dict.h"
+#include "arrow/array/builder_primitive.h"
+#include "arrow/array/dict_internal.h"
+#include "arrow/compute/api_vector.h"
+#include "arrow/compute/kernels/common_internal.h"
+#include "arrow/result.h"
+
+namespace arrow {
+
+namespace compute {
+namespace internal {
+
+namespace {
+
+// Dictionary compaction implementation
+
+const FunctionDoc dictionary_compact_doc{
+    "Compact dictionary array",
+    ("Return a compacted version of the dictionary array input,\n"
+     "which removes unused values in dictionary.\n"
+     "The function assumes every indices has its corresponding value\n"
+     "in dictionary."),
+    {"dictionary_array"}};
+
+class DictionaryCompactKernel : public KernelState {
+ public:
+  virtual Result<std::shared_ptr<Array>> Exec(std::shared_ptr<Array> dict_array,
+                                              ExecContext* ctx) const = 0;
+};
```

Comment:
Hi, I change to `Exec(KernelContext* ctx, const ExecSpan& batch, ExecResult* out)`. Does it ok? Since there is error when I use `ArrayKernelExec exec =  Exec` in DictionaryCompactKernelImpl<T>.

One curious question, for the old version code, the data is warpped by `std::shared_ptr<Array>`. Since it's a pointer, the performance could be Ok? Is there some subtle places to influence the performance?